wagtail/core/collectors.py

```diff
@@ -0,0 +1,379 @@
+            values = []
+            for obj in searched_objects:
+                for model, handler in handlers.items():
+                    if isinstance(obj, model):
+                        k, v = handler.get_id_pair_from_instance(obj)
+                        values.append('%s="%s"' % (k, re.escape(str(v))))
+            params['val'] = (values[0] if len(values) == 1
+                             else r'(%s)' % r'|'.join(values))
+        else:
+            pattern = r'<(a|embed) %(type)s[^>]*>'
+        return pattern.replace(r' ', r'[^>]*[ \t\n]') % params
+
+    def find_objects(self, *searched_objects):
+        if not self.fields:
+            return
+
+        pattern = self.get_pattern_for_objects(searched_objects)
+        if pattern is None:
+            return
+
+        filters = Q()
+        for field in self.fields:
+            filters |= Q(**{field.attname + '__regex': pattern})
+
+        searched_data = {get_obj_base_key(obj) for obj in searched_objects}
+        for obj in self.model._default_manager.using(self.using) \
+                .filter(filters):
+            for field in self.fields:
+                for found_obj in find_objects_in_rich_text(
+                        getattr(obj, field.attname)):
+                    if not searched_objects or \
+                            get_obj_base_key(found_obj) in searched_data:
+                        yield obj, found_obj
+
+
+class StreamFieldCollector:
+    def __init__(self, field):
+        self.field = field
+
+    def block_tree_paths(self, block, ancestors=()):
+        if isinstance(block, (StreamBlock, StructBlock)):
+            for child_block in block.child_blocks.values():
+                yield from self.block_tree_paths(child_block,
+                                                 ancestors + (block,))
+        elif isinstance(block, ListBlock):
+            yield from self.block_tree_paths(block.child_block,
+                                             ancestors + (block,))
+        else:
+            yield ancestors + (block,)
+
+    def find_block_type(self, block_types):
+        for block_path in self.block_tree_paths(self.field.stream_block):
+            if isinstance(block_path[-1], block_types):
+                yield block_path
+
+    def find_values(self, stream, block_path):
+        if not block_path:
+            if isinstance(stream, RichText):
+                yield from find_objects_in_rich_text(stream.source)
+            else:
+                yield stream
+            return
+        current_block, *block_path = block_path
+        if isinstance(current_block, StreamBlock) \
+                and isinstance(stream, StreamValue):
+            for sub_value in stream:
+                yield from self.find_values(sub_value, block_path)
+        elif isinstance(stream, StreamValue.StreamChild):
+            if stream.block == current_block:
+                yield from self.find_values(stream.value, block_path)
```

Comment:
These recursion rules don't seem quite right: unwrapping `StreamValue` and `StreamChild` happens over two iterations of `find_values`, during which two blocks will be popped off `block_path` - but both wrapper elements correspond to a `StreamBlock`, so logically only one block should have been popped by that point.

(Looking closer, I think there's some inconsistency about whether the type of `stream` corresponds to the first block in `block_path`, or the block that's just been popped in the previous iteration, and my guess is that this inconsistency only surfaces in nested `StreamBlock`s, which aren't covered by the tests.)

Reply:
I’m not sure I fully follow you here.

I started building this recursive method using my knowledge of the way `StructBlock`, `ListBlock` and `StreamBlock` work, but it ended up being more complex, with several intermediate classes (that could probably be refactored). I adjusted the method to work in all the *scenarii* I could think of. Could you make an example that would not be handled by this?

Reply:
Failing test (throwing an 'Unexpected StreamField value' warning) here: https://github.com/gasman/wagtail/commit/4ed64fa82c72bebfb089a39e2298ebe1bbc5412e

You can verify that there's a logic error here by asking: does the value of `stream` correspond to the first item of `block_path`, or not?

* The line `if isinstance(current_block, StreamBlock) and isinstance(stream, StreamValue):` suggests that it _does_ correspond
* The `if not block_path:` section at the top suggests that it _doesn't_ correspond (`block_path` is empty, but we still have a `stream` value)
* `elif current_block.name == '' and isinstance(stream, list):` also suggests that it _doesn't_ correspond (the 'nameless' block in a ListBlock definition is the child, not the ListBlock itself)

Unfortunately at this point I really have to ask you to add docstrings to the remainder of collectors.py before I can continue reviewing. Throughout this code there's a lack of precision about the inputs / outputs of each method... up to now I was willing to approach that as a documentation issue, and fill in that missing detail as part of the review process. However, it now seems that this lack of precision is directly causing logical errors such as this one and the regexp escaping bug earlier - at this point, I don't trust my ability to catch these logical errors at the same time as figuring out the intended behaviour of the code.